crates/containerd-shim-wasm/src/sandbox/oci.rs

```diff
@@ -44,8 +29,51 @@ pub fn get_args(spec: &Spec) -> &[String] {
     }
 }
 
-pub fn spec_from_file<P: AsRef<Path>>(path: P) -> Result<Spec, Error> {
+pub fn spec_from_file<P: AsRef<Path>>(path: P) -> Result<Spec> {
     let file = File::open(path)?;
     let cfg: Spec = json::from_reader(file)?;
     Ok(cfg)
 }
+
+struct NopCgroup {}
+
+impl cgroups::Cgroup for NopCgroup {
+    fn add_task(&self, _pid: u32) -> Result<()> {
+        Ok(())
+    }
+
+    fn version(&self) -> cgroups::Version {
+        cgroups::Version::V1
+    }
+
+    fn apply(&self, _res: Option<cgroups::Resources>) -> Result<()> {
+        Ok(())
+    }
+
+    fn delete(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub fn get_cgroup(spec: &Spec) -> Result<Box<dyn cgroups::Cgroup>> {
+    let linux = spec.linux();
+    if linux.is_none() {
+        return Ok(Box::new(NopCgroup {}));
+    }
+
+    let p = linux.as_ref().unwrap().cgroups_path();
+    if p.is_none() {
+        return Ok(Box::new(NopCgroup {}));
+    }
+
+    cgroups::new(p.as_ref().unwrap().to_str().unwrap().to_string())
+}
+
+pub fn setup_cgroup(cg: &Box<dyn cgroups::Cgroup>, spec: &Spec) -> Result<()> {
```

Comment:
```suggestion
pub fn setup_cgroup(cg: &dyn cgroups::Cgroup, spec: &Spec) -> Result<()> {
```

~don't need `Box` if we are passing a reference, will suggest fixes on the function calls of this too.